lib/pbench/server/filetree.py

```diff
@@ -299,18 +300,27 @@ def unpack(self, incoming: Path, results: Path):
         ensured by calling this indirectly through the Controller class unpack
         method.
 
-        TODO: This is a prototype for testing, and doesn't actually unpack
-        the tarball as we're going to be relying on the current unpack pipeline
-        command for some time.
-
         Args:
             incoming: Controller's directory in the INCOMING tree
             results: Controller's directory in the RESULTS tree
         """
-        self.logger.warning("Tarball unpack is not yet implemented")
         unpacked = incoming / self.name
-        unpacked.mkdir(parents=True)  # Just create an empty directory for now
+        unpacked.mkdir(parents=True)
         self.unpacked = unpacked
+
+        def unpack_tarball(tarball_path, unpacked_dir):
+            # extracting Tarball in unpacked directory
+            tarfile.open(tarball_path).extractall(unpacked_dir)
+            return
+
+        p = mp.Process(target=unpack_tarball, args=(self.tarball_path, self.unpacked))
+        p.start()
+
+        if p.exitcode != 0:
```

Comment:
If we have a failure extracting a tar ball, do we want to continue performing the steps at line 324?  Or should be we consider exiting without changing the state?  Or should we move this to a quarantine state?

Reply:
I think we should move it to `WONT-UNPACK` state. What do you think @dbutenhof

Reply:
I'm in the process of removing "state" directory links; and since it turns out that nearly all our use of "quarantine" is currently for state links that don't resolve, I'm pretty much trashing that as well in favor of simply tracking errors as dataset metadata. (Eventually we'll have the audit log.)

The `FileTree` currently recognizes the state directories, but isn't driven by them. I haven't really figured out how your changes to `pbench-unpack-tarball.sh` and mine will interleave. For now, I like that you've implemented the code here but haven't changed the current unpack code. To activate this we need to either rewrite the unpacker as Python (not a bad thing) or use something like `pbench-tree-manage --unpack` (which I didn't implement yet). Meanwhile, I'm tearing out the use of state directory links.

In any case, I don't think the code *here* should be doing anything; this is a low level utility method. It reports failure, and I suspect shouldn't set the `self.unpacked` link, on failure. (That is, unpack to `unpacked` and then set `self.unpacked = unpack` only on success.) So I suspect we should be raising something like `TarballUnpackError` with a description, and let the caller figure out what to do about it.

In terms of `Dataset` states, there no longer is an `UNPACKED` state; it'll just say `UPLOADED` and we'll log the error. (With my changes, `quarantine` becomes a method to add `server.errors` metadata recording the failure, so we'll have something like

```
"server": {
  "errors": {
    "unpack": "Not a tar file"
  }
}
```

Reply:
Latest changes address this concern well.